src/Database/PG/Query/Connection.hs

```diff
@@ -365,20 +369,48 @@ execPrepared
   :: PQ.Connection                -- ^ connection
   -> [Maybe (DB.ByteString, PQ.Format)]           -- ^ parameters
   -> RemoteKey                    -- ^ stmtName
+  -> ((PQ.Connection -> IO (Maybe PQ.Result)) -> PQ.Connection -> IO (Maybe PQ.Result))
   -> PGExec ResultOk -- ^ result
-execPrepared conn args n = do
-  mRes <- lift $ PQ.execPrepared conn n args PQ.Binary
+execPrepared conn args n run = do
+  mRes <- lift $ run (\c -> PQ.execPrepared c n args PQ.Binary) conn
   checkResult conn mRes
 
+data PGCancelErr = CEAllocate | CEError DT.Text
+  deriving (Show, Eq)
+
+instance Exception PGCancelErr
+
+cancelPG :: PQ.Connection -> IO (Either PGCancelErr ())
+cancelPG conn = do
+  PQ.getCancel conn >>= \case
+    Nothing -> return $ Left CEAllocate
+    Just c  -> PQ.cancel c >>= \case
+      Left err -> return $ Left $ CEError $ lenientDecodeUtf8 err
+      Right () -> return $ Right ()
+
+cancelOnTimeout :: (PQ.Connection -> IO a) -> PQ.Connection -> IO a
+cancelOnTimeout f conn = do
+  a <- async $ f conn
+  x <- (Left <$> wait a) `catch`
+    (\(e::Timeout) -> Right <$> do
+        cancelPG conn >>= \case
+          Left err -> return $ Left (err, e)
+          Right () -> return $ Right ())
+  case x of
+    Left res               -> return res
+    Right (Left (err, _e)) -> throw err -- very unclear this is correct, just want to avoid losing the error for now
```

Comment:
You're gonna want `throwIO` instead of `throw` here; `throw :: Exception e => e -> a` throws the exception in a pure context which can be a pain to deal with. It _shouldn't_ be a problem in this case, but it's a good pattern to avoid.